rent/src/lib.rs

```diff
@@ -0,0 +1,1611 @@
+		/// cancellation fee.
+		FlexibleFeeOnlyForFixedDuration,
+		/// Cannot create a contract with no revocation type and a renter cancellation fee.
+		NoRenterCancellationFeeWithNoRevocation,
+		/// Cannot create a contract with NFT id rent fee type and subscription duration.
+		NoNFTRentFeeWithSubscription,
+		/// The contract was not found for the given nft_id.
+		ContractNotFound,
+		/// The caller is neither the renter or rentee.
+		NotTheRenterOrRentee,
+		/// The caller is not the renter.
+		NotTheRenter,
+		/// The caller is not the rentee.
+		NotTheRentee,
+		/// Operation is not permitted because revocation type is not anytime.
+		CannotRevoke,
+		/// End block not found for flexible fee calculation.
+		FlexibleFeeEndBlockNotFound,
+		/// Cannot Rent your own contract.
+		CannotRentOwnContract,
+		/// Rentee is not on authorized account list.
+		NotAuthorizedForRent,
+		/// Operation is not permitted because user has not enough funds.
+		NotEnoughBalance,
+		/// Operation is not permitted because user has not enough funds for cancellation fee.
+		NotEnoughBalanceForCancellationFee,
+		/// Operation is not permitted because user has not enough funds for rent fee.
+		NotEnoughBalanceForRentFee,
+		/// Fee NFT not found.
+		NFTNotFoundForRentFee,
+		/// Maximum offers reached.
+		MaximumOffersReached,
+		/// Operation is not permitted because acceptance type is not manual.
+		CannotAcceptOfferForAutoAcceptance,
+		/// Operation is not permitted because acceptance type is not manual.
+		CannotRetractOfferForAutoAcceptance,
+		/// Operation is not permitted because contract has started.
+		ContractHasStarted,
+		/// Operation is not permitted because contract has not started yet.
+		ContractHasNotStarted,
+		/// Terms can only be set for subscription duration and OnSubscriptionChange revocation
+		/// type.
+		CanChangeTermForSubscriptionOnly,
+		/// New term must be suvscription.
+		CanSetTermsForSubscriptionOnly,
+		/// Operation is not permitted because contract terms are already accepted
+		ContractTermsAlreadyAccepted,
+		/// Operation is allowed only for subscription
+		RenewalOnlyForSubscription,
+		/// Operation is not allowed because same nft was used for contract /
+		/// renter_cancellation_fee / rentee_cancellation_fee / rent_fee
+		InvalidFeeNFT,
+		/// No offers was found for the contract
+		NoOffersForThisContract,
+		/// No offer was made by the rentee
+		NoOfferFromRentee,
+		/// Math error.
+		InternalMathError,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Weight: see `begin_block`
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let mut read = 0;
+			let mut write = 0;
+			let mut current_actions = 0;
+			let max_actions = T::ActionsInBlockLimit::get();
+
+			// Fixed queue management
+			while current_actions < max_actions {
+				let fixed_queue = FixedQueue::<T>::get();
+				read += 1;
+
+				if let Some(nft_id) = fixed_queue.next(now) {
```

Comment:
This is a huge performance issue. If we need to end 1k  contracts in one block we are going to have at least 1k calls to change the `FixedQueue` storage.